### PR TITLE
Environment Enum

### PR DESCRIPTION
**Pull request by mattbriancon**
I needed a consistent way to get the short and long environment names in mbq.pubsub and I figured that'd be useful everywhere. This solution was the first thing that came to mind; very open to feedback.

## Features

```python
[I] matt@matt ~/P/q/mbq.env> ipython
Python 3.6.3 (default, Dec  8 2017, 14:59:20)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from mbq import env

In [2]: env.Environment.PRODUCTION.short_name
Out[2]: 'prd'

In [3]: env.Environment.PRODUCTION.long_name
Out[3]: 'Production'

In [4]: env.Environment.PRODUCTION.is_deployed
Out[4]: True

In [5]: prd = env.Environment('prd')

In [6]: bad = env.Environment('bad')
# ...snip...
ValueError: 'bad' is not a valid Environment

In [7]: prd is env.Environment.PRODUCTION
Out[7]: True

In [8]: import os

In [8]: os.environ['ENV_NAME'] = 'pRoDuCtIoN'

In [9]: env.get_environment('ENV_NAME')
Out[9]: <Environment.PRODUCTION: ('prd', 'Production')>

In [10]: env.get_environment('ENV_NAME') is env.Environment.PRODUCTION
Out[10]: True
```

### CODE REVIEW

File: mbq/env/__init__.py [hunk 1]
```diff
@@ -7,30 +9,73 @@ class EnvException(Exception):
...
+    represent the environment in which a service is running.
+    """
+
+    PRODUCTION = ('prd', 'Production')
+    DEVELOPMENT = ('dev', 'Development')
+    LOCAL = ('lcl', 'Local')
+
+    def __init__(self, short, long):
+        self.short = short
+        self.long = long
```
Comment (thieman) [inline mbq/env/__init__.py]: Is `long` a keyword? GitHub's highlighting it

File: mbq/env/__init__.py [hunk 2]
```diff
@@ -7,30 +9,73 @@ class EnvException(Exception):
...
-        except KeyError:
+        except KeyError as e:
             if default is not NOT_PROVIDED:
                 return default
 
             if not required:
                 return None
 
-            raise EnvException('Missing key "{}"'.format(key))
+            raise EnvException('Missing key "{}"'.format(key)) from e
```
Comment (thieman) [inline mbq/env/__init__.py:57]: Whoa TIL

File: mbq/env/__init__.py [hunk 3]
```diff
@@ -7,30 +9,73 @@ class EnvException(Exception):
...
         if coerce is not NOT_PROVIDED:
             val = coerce(val)
 
         return val
 
+    def get_environment(self, key, default=NOT_PROVIDED, required=True):
+        try:
+            return self.get(key, default=default, required=required, coerce=Environment.from_value)
+        except ValueError as e:
+            if 'is not a valid Environment' in str(e):
```
Comment (thieman) [inline mbq/env/__init__.py]: This is crazy gross, why do we need custom handling here at all?

File: tests/test_env.py [hunk 4]
```diff
@@ -116,6 +116,13 @@ def test_passes_stripped_val_to_coerce(self):
         args, kwargs = passthru.call_args
         self.assertEqual(args[0], 'value')
 
+    @environ(GOOD='prd', BAD='BACON')
```
Comment (thieman) [inline tests/test_env.py:119]: Some questionable judgments being made here w.r.t. the quality of bacon

File: mbq/env/__init__.py (hunk 1 above)
Comment (mattbriancon) [reply mbq/env/__init__.py]: It was in Py2

File: mbq/env/__init__.py (hunk 2 above)
Comment (mattbriancon) [reply mbq/env/__init__.py:57]: Yeah it's a minor thing to make stack trace a bit easier to comprehend

File: mbq/env/__init__.py (hunk 3 above)
Comment (mattbriancon) [reply mbq/env/__init__.py]: Yeah I went back and forth here. It's mostly to keep the API clean so you can always expect an `EnvException` (see `get_int`). Down to remove it but then let's remove it in `get_int` too.

Comment (thieman) [reply mbq/env/__init__.py]: If the intent is to always return an `EnvException` then why do we need to branch on the type of error?

Comment (mattbriancon) [reply mbq/env/__init__.py]: Granted, exceptions can be raised from anywhere and we're not wrapping everything in a `try/except` but where we know certain exceptions are likely or even intentional, converting the exception to something expected can make interacting with the library easier.

```python
try:
    env.get_environment('ENV_NAME')
except env.EnvException as e:
    # handle everything related to mbq.env
except ValueError:
    # figure out if this is the ValueError from the Enum or something else?
```

Comment (mattbriancon) [reply mbq/env/__init__.py]: Changed it to let the ValueError through. I don't feel strongly about it and the traceback should make it fairly straightforward to figure out.

**Comment by mattbriancon** [comment]
Fixed up a couple things, made the short and long attrs short_name and long_name, and made the lookup interface feel a bit more native. Merging so I can keep moving.

**Pull request merged**